File: Top-Interview-Questions-Leetcode/Easy-Collections/strings/implement-strStr-function.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        # return haystack.find(needle)
        lh, ln = len(haystack), len(needle)
        
        for i in range(lh - ln + 1):
            if haystack[i: i + ln] == needle: 
                return i
        
        return -1

# Example usage
def main():
    haystack = "dangphanthanh"
    needle = "hanh"
    solution = Solution()
    result = solution.strStr(haystack, needle)
    print(result) 

if __name__ == "__main__":
    main()

File: Top-Interview-Questions-Leetcode/Easy-Collections/strings/implement-strStr-function.cpp
#include <iostream>
#include <string>

class Solution {
public:
    int strStr(std::string haystack, std::string needle) {
        // if (haystack == needle) return 0;
        
        int nh = haystack.size(), nn = needle.size();
        
        for (int i = 0; i <= nh - nn; ++i) {
            int j = 0, i_temp = i;
            
            while (j < nn and haystack[i_temp] == needle[j]) i_temp++, j++;
            
            if (j >= nn) return i;
        }
        
        return -1;
    }
};

// Example usage
int main() {
    Solution solution;
    std::string haystack = "phanthanhdang", needle = "hanh";
    int result = solution.strStr(haystack, needle);
    std::cout << "The index of the first occurrence of needle in haystack is: " << result << std::endl;
    return 0;
}